File: ios/Services/KitharaPlayer/Sources/ServicesKitharaPlayer/CompositePlayableUrlResolver.swift
import CoreDomain
import Foundation

public struct CompositePlayableUrlResolver: PlayableUrlResolver, Sendable {
    private let providers: [MusicProviderId: TrackStreamUrlProvider]

    public init(providers: [MusicProviderId: TrackStreamUrlProvider]) {
        self.providers = providers
    }

    public func playableURL(for item: PlaybackQueueItemDescriptor) async throws -> URL {
        switch item.source {
        case .local(let url):
            return url

        case .remote(let provider):
            guard let provider = providers[provider] else {
                throw MusicLibraryError.providerError(
                    code: "missing-provider",
                    description: "No stream URL provider registered for \(provider)."
                )
            }
            return try await provider.streamURL(for: item.trackId)
        }
    }
}

struct ProviderPlayableUrlResolver: PlayableUrlResolver {
    let urlProvider: TrackStreamUrlProvider

    func playableURL(for item: PlaybackQueueItemDescriptor) async throws -> URL {
        switch item.source {
        case .local(let url):
            return url

        case .remote:
            return try await urlProvider.streamURL(for: item.trackId)
        }
    }
}

File: android/services/kithara/src/main/java/com/mplayeraudio/services/kithara/CompositePlayableUrlResolver.kt
package com.mplayeraudio.services.kithara

import com.mplayeraudio.core.domain.musiclibrary.MusicLibraryException
import com.mplayeraudio.core.domain.musiclibrary.MusicProviderId
import com.mplayeraudio.core.domain.musiclibrary.TrackStreamUrlProvider
import com.mplayeraudio.core.player.PlayableSource
import com.mplayeraudio.core.player.PlayableUrlResolver
import com.mplayeraudio.core.player.PlaybackQueueItem

internal class CompositePlayableUrlResolver(
    private val providers: Map<MusicProviderId, TrackStreamUrlProvider>,
) : PlayableUrlResolver {

    override suspend fun getPlayableUrl(item: PlaybackQueueItem): String {
        return when (val source = item.source) {
            is PlayableSource.Local -> source.uri
            is PlayableSource.Remote -> provider(source.provider).getStreamUrl(item.trackId)
        }
    }

    private fun provider(providerId: MusicProviderId): TrackStreamUrlProvider {
        return providers[providerId]
            ?: throw MusicLibraryException.ProviderError(
                code = "missing-provider",
                description = "No stream URL provider registered for $providerId.",
            )
    }
}
